webui/react/src/pages/Authentication.tsx

```diff
@@ -0,0 +1,156 @@
+import Spinner from 'components/Spinner';
+import Auth, { updateAuth } from 'contexts/Auth';
+import handleError, { ErrorType } from 'ErrorHandler';
+import { crossoverRoute, isCrossoverRoute } from 'routes';
+import { isLoginFailure, login, logout } from 'services/api';
+import { Credentials } from 'types';
+
+import css from './Authentication.module.scss';
+
+const DEFAULT_REDIRECT = '/det/dashboard';
+
+type WithSearch<T> = T & {location: {search: string}};
+interface Queries {
+  redirect?: string;
+}
+
+interface FromValues {
+  password?: string;
+  username?: string;
+}
+
+const Authentication: React.FC<WithSearch<{}>> = (props: WithSearch<{}>) => {
+  const history = useHistory();
+  const location = useLocation();
+  const auth = Auth.useStateContext();
+  const setAuth = Auth.useActionContext();
+  const [ isLoading, setIsLoading ] = useState(true);
+  const [ badCredentials, setBadCredentials ] = useState(false);
+  const [ canSubmit, setCanSubmit ] = useState(false);
+
+  const queries: Queries = queryString.parse(props.location.search);
+
+  const redirectUri= queries.redirect || DEFAULT_REDIRECT;
+
+  useEffect(() => {
+    const source = axios.CancelToken.source();
+    updateAuth(setAuth, source.token).then(() => setIsLoading(false));
+    return (): void => {
+      source.cancel();
+    };
+  }, [ setAuth ]);
+
+  const isLogout = location.pathname.endsWith('logout');
+  if (isLogout) {
+    logout({});
+    setAuth({ type: Auth.ActionType.Reset });
+    history.push('/det/login' + props.location.search);
+  }
+
+  const onFinish = (creds: FromValues): void => {
+    // TODO validate the creds type?
+    setCanSubmit(false);
+    const hideLoader = message.loading('logging in..');
+    login(creds as Credentials)
+      .then(() => updateAuth(setAuth))
+      .catch((e: Error) => {
+        setBadCredentials(isLoginFailure(e));
+        if (badCredentials) return;
+        // DISCUSS do we want to potentially report it?
+        // or pass it through error handler for some reason?
+        const actionMsg = badCredentials ? 'check your username and password.' : 'retry.';
+        handleError({
+          error: e,
+          isUserTriggered: true,
+          message: e.message,
+          publicMessage: `Failed to login. Please ${actionMsg}`,
+          publicSubject: 'Login failed',
+          type: badCredentials ? ErrorType.Input : ErrorType.Server,
+        });
+      })
+      .finally(() => {
+        setCanSubmit(true);
+        hideLoader();
+      });
+  };
+
+  const onValuesChange = (changes: FromValues, values: FromValues): void => {
+    const hasUsername = !!values.username;
+    setBadCredentials(false);
+    setCanSubmit(hasUsername);
+  };
+
+  const loginForm = (
+    <Form
+      name="basic"
+      size="large"
+      onFinish={onFinish}
+      onValuesChange={onValuesChange}
+    >
+      <Form.Item
+        name="username"
+        rules={[
+          {
+            message: 'Please input your username!',
```

Comment:
How about we remove the exclamation? I feel like I'm being badgered :)
`Please type in your username.` feels more human-friendly, wouldn't you agree?

Reply:
I agree : )

Reply:
`username input does not compute!!!`